src/freenas/usr/local/lib/middlewared_truenas/plugins/failover_/remote.py

```diff
@@ -0,0 +1,255 @@
+                refused = True
+            except Exception:
+                logger.error('Remote connection failed', exc_info=True)
+                refused = False
+            time.sleep(retry)
+
+    def connect_and_wait(self):
+        try:
+            # 860 is the iSCSI port and blocked by the failover script
+            with Client(
+                f'ws://{self.remote_ip}:6000/websocket',
+                reserved_ports=True, reserved_ports_blacklist=[860],
+            ) as c:
+                self.client = c
+                self.connected.set()
+                # Subscribe to all events on connection
+                with self._subscribe_lock:
+                    for name in self._subscriptions:
+                        self.client.subscribe(name, partial(self._sub_callback, name))
+                self._on_connect()
+                c._closed.wait()
+        except OSError as e:
+            if e.errno in (
+                errno.EPIPE,  # Happens when failover is configured on cxl device that has no link
+                errno.ENETDOWN, errno.EHOSTDOWN, errno.ENETUNREACH, errno.EHOSTUNREACH,
+                errno.ECONNREFUSED,
+            ) or isinstance(e, socket.timeout):
+                raise ConnectionRefusedError()
+            raise
+        finally:
+            if self.connected.is_set():
+                # Only happens if we have successfully connected once
+                self._on_disconnect()
+            self.client = None
+            self.connected.clear()
+
+    def is_connected(self):
+        return self.connected.is_set()
+
+    def register_connect(self, cb):
+        """
+        Register a callback to be called everytime we connect to the other node.
+        """
+        self._on_connect_callbacks.append(cb)
+
+    def _on_connect(self):
+        """
+        Called everytime connection has been established.
+        """
+        for cb in self._on_connect_callbacks:
+            try:
+                cb(self.middleware)
+            except Exception:
+                logger.error('Failed to run on_connect for remote client', exc_info=True)
+
+    def register_disconnect(self, cb):
+        """
+        Register a callback to be called everytime we disconnect from the other node.
+        """
+        self._on_disconnect_callbacks.append(cb)
+
+    def _on_disconnect(self):
+        """
+        Called everytime connection is closed for whatever reason.
+        """
+        for cb in self._on_disconnect_callbacks:
+            try:
+                cb(self.middleware)
+            except Exception:
+                logger.error('Failed to run on_disconnect for remote client', exc_info=True)
+
+    def call(self, *args, **kwargs):
+        try:
+            if not self.connected.wait(timeout=20):
+                if self.remote_ip is None:
+                    raise CallError('Unable to determine remote node IP')
```

Comment:
How about we store that exception in Client class and raise it? So we can see why we are unable to determine remote node IP.

Reply:
There is no reason why. It just means it couldnt detect A or B